pkg/metrics/providers/datadog.go

```diff
@@ -0,0 +1,160 @@
+
+	fromDeltaMultiplierOnProgressiveDeadline = 10
+)
+
+// DatadogProvider executes datadog queries
+type DatadogProvider struct {
+	metricsQueryEndpoint     string
+	apiKeyValidationEndpoint string
+
+	timeout        time.Duration
+	apiKey         string
+	applicationKey string
+	fromDelta      int64
+}
+
+type datadogResponse struct {
+	Series []struct {
+		Pointlist [][]float64 `json:"pointlist"`
+	}
+}
+
+// NewDatadogProvider takes a canary spec, a provider spec and the credentials map, and
+// returns a Datadog client ready to execute queries against the API
+func NewDatadogProvider(cd *flaggerv1.Canary,
+	provider flaggerv1.MetricTemplateProvider,
+	credentials map[string][]byte) (*DatadogProvider, error) {
+
+	address := provider.Address
+	if address == "" {
+		address = datadogDefaultHost
+	}
+
+	dd := DatadogProvider{
+		timeout:                  5 * time.Second,
+		metricsQueryEndpoint:     address + datadogMetricsQueryPath,
+		apiKeyValidationEndpoint: address + datadogAPIKeyValidationPath,
+	}
+
+	if b, ok := credentials[datadogAPIKeySecretKey]; ok {
+		dd.apiKey = string(b)
+	} else {
+		return nil, fmt.Errorf("datadog credentials does not contain datadog_api_key")
+	}
+
+	if b, ok := credentials[datadogApplicationKeySecretKey]; ok {
+		dd.applicationKey = string(b)
+	} else {
+		return nil, fmt.Errorf("datadog credentials does not contain datadog_application_key")
+	}
+
+	dd.fromDelta = int64(cd.GetProgressDeadlineSeconds() * fromDeltaMultiplierOnProgressiveDeadline)
+	return &dd, nil
+}
+
+// RunQuery executes the datadog query against DatadogProvider.metricsQueryEndpoint
+// and returns the the first result as float64
+func (p *DatadogProvider) RunQuery(query string) (float64, error) {
+
+	req, err := http.NewRequest("GET", p.metricsQueryEndpoint, nil)
+	if err != nil {
+		return 0, fmt.Errorf("error http.NewRequest: %s", err.Error())
+	}
+
+	req.Header.Set(datadogAPIKeyHeaderKey, p.apiKey)
+	req.Header.Set(datadogApplicationKeyHeaderKey, p.applicationKey)
+
+	q := req.URL.Query()
+
+	now := time.Now().Unix()
+	q.Add("query", query)
+	q.Add("from", strconv.FormatInt(now-p.fromDelta, 10))
+	q.Add("to", strconv.FormatInt(now, 10))
+	req.URL.RawQuery = q.Encode()
+
+	ctx, cancel := context.WithTimeout(req.Context(), p.timeout)
+	defer cancel()
+	r, err := http.DefaultClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return 0, err
+	}
+
+	defer r.Body.Close()
+	b, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return 0, fmt.Errorf("error reading body: %s", err.Error())
+	}
+
+	if r.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("error response: %s", string(b))
+	}
+
+	var res datadogResponse
+	if err := json.Unmarshal(b, &res); err != nil {
+		return 0, fmt.Errorf("error unmarshaling result: %s, '%s'", err.Error(), string(b))
+	}
+
+	if len(res.Series) < 1 {
+		return 0, fmt.Errorf("error getting times series: %s", string(b))
+	}
+
+	s := res.Series[0]
+	return s.Pointlist[len(s.Pointlist)-1][1], nil
```

Comment:
We should check the length of point list and return `no values found` before trying to get the value.